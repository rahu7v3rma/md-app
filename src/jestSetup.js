import mockRNCameraRoll from '@react-native-camera-roll/camera-roll/src/__mocks__/nativeInterface';
import mockRNCNetInfo from '@react-native-community/netinfo/jest/netinfo-mock';
import mockRNDeviceInfo from 'react-native-device-info/jest/react-native-device-info-mock';

import { MockData } from './utils/mockData';

// setImmediate and clearImmediate aren't mocked by useFakeTimers if they are
// not set to a function (https://github.com/jestjs/jest/pull/11599)
global.setImmediate = jest.fn();
global.clearImmediate = jest.fn();
jest.useRealTimers();

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

jest.mock('react-native-keychain', () => {
    return {
        SECURITY_LEVEL_ANY: 'MOCK_SECURITY_LEVEL_ANY',
        SECURITY_LEVEL_SECURE_SOFTWARE: 'MOCK_SECURITY_LEVEL_SECURE_SOFTWARE',
        SECURITY_LEVEL_SECURE_HARDWARE: 'MOCK_SECURITY_LEVEL_SECURE_HARDWARE',
        setGenericPassword: jest.fn().mockResolvedValue(),
        getGenericPassword: jest.fn().mockResolvedValue(false),
        resetGenericPassword: jest.fn().mockResolvedValue()
    };
});

jest.mock('react-native-splash-screen', () => {
    return {
        hide: jest.fn()
    };
});

export let mockUnwrap = jest.fn().mockImplementation(() => {
    return Promise.resolve({});
});
export const mockDispatch = jest.fn().mockImplementation(() => {
    return {
        unwrap: mockUnwrap
    };
});
jest.mock('@/hooks', () => ({
    useAppDispatch: () => mockDispatch,
    useChannelPreviewInfo: () => mockDispatch
}));
export const mockChannelId = 'mockChannelId';
export const mockChatClient = jest.fn().mockImplementation(() => ({
    userID: 1,
    chat_id: 1,
    queryChannels: jest.fn().mockImplementation(() => {
        return Promise.resolve([
            {
                cid: 99
            }
        ]);
    }),
    activeChannels: {
        mockChannelId
    }
}));
jest.mock('@/contexts/appChat', () => ({
    useAppChat: () => {
        return {
            chatClient: mockChatClient(),
            setActiveChatChannel: jest.fn()
        };
    }
}));
export const mockRNSha256 = jest
    .fn()
    .mockImplementation(() => Promise.resolve('123'));
jest.mock('react-native-sha256', () => {
    return {
        sha256: mockRNSha256
    };
});

jest.mock('react-native-share', () => ({
    default: jest.fn()
}));

jest.mock('@/services/api', () => ({
    getImageUploadCredentials: jest.fn(() => ({
        keyPrefix: 'keyPrefix',
        fileName: 'fileName',
        bucketName: 'bucketName'
    }))
}));

jest.mock('@/reducers/content', () => ({
    ContentSelectors: jest.fn().mockImplementation(() => ({
        loading: false,
        journey: MockData.journey,
        navigationKey: 'Main.Home',
        activeBlock: MockData.activeBlock,
        activeLesson: {
            lesson: {}
        },
        journeyBlocks: MockData.journeyBlocks
    })),
    loadJourney: jest.fn(),
    getActiveLesson: jest.fn()
}));

export const mockChatProfile = jest.fn(() => ({
    apiKey: 'apiKey',
    token: 'token',
    userId: 'userId',
    firebasePushProviderName: 'firebasePushProviderName'
}));
export const mockDeleteAccount = jest.fn();
jest.mock('@/reducers/user', () => ({
    UserSelectors: jest.fn().mockImplementation(() => ({
        loading: false,
        userInfo: {
            username: 'user1',
            first_name: 'Some',
            last_name: 'Name',
            email: 'user1@example.example'
        },
        userProfile: {
            diabetes_type: 'Prediabetes',
            image: null,
            onboarding_form_url: null
        },
        coach: {
            first_name: 'name',
            profile_image: 'profile_image'
        },
        chatProfile: mockChatProfile(),
        hasCoachChat: false
    })),
    userLogin: jest.fn(),
    changePassword: jest.fn(),
    resetPassword: jest.fn(),
    resetPasswordConfirm: jest.fn(),
    resetPasswordVerify: jest.fn(),
    getCoach: jest.fn(),
    refreshProfileSession: jest.fn(),
    getProfile: jest.fn(),
    searchFilterAction: jest.fn(),
    clearSearchText: jest.fn(),
    deleteAccount: mockDeleteAccount
}));
export const mockRNFSExists = jest
    .fn()
    .mockImplementation(() => Promise.resolve(false));
export const mockRNFSCachesDirectoryPath = 'mockCachesDirectoryPath';
export const mockRNFSDownloadFile = jest.fn().mockImplementation(() => ({
    promise: Promise.resolve({ statusCode: 200 })
}));
jest.mock('react-native-fs', () => {
    return {
        readFile: jest.fn(),
        exists: mockRNFSExists,
        CachesDirectoryPath: mockRNFSCachesDirectoryPath,
        downloadFile: mockRNFSDownloadFile,
        unlink: jest.fn().mockImplementation(() => Promise.resolve())
    };
});
jest.mock('react-native-image-crop-picker', () => {
    return {
        openPicker: jest.fn(),
        launchImageLibrary: jest.fn(),
        openCamera: jest.fn()
    };
});

jest.mock('@react-native-camera-roll/camera-roll', () => mockRNCameraRoll);
jest.mock('react-native-reanimated', () => ({
    ...require('react-native-reanimated/mock')
}));
jest.mock('react-native-device-info', () => mockRNDeviceInfo);
jest.mock('@react-native-async-storage/async-storage', () =>
    require('@react-native-async-storage/async-storage/jest/async-storage-mock')
);
jest.mock('react-native/Libraries/EventEmitter/NativeEventEmitter');

jest.mock('@/reducers/track', () => ({
    TrackSelectors: jest.fn().mockImplementation(() => ({
        trackFastState: '',
        trackStartedAt: '',
        trackTimeLimitInSeconds: 57600,
        trackedTimeInSeconds: 0
    })),
    trackStart: jest.fn(),
    trackUpdate: jest.fn()
}));

jest.mock('@react-native-community/netinfo', () => mockRNCNetInfo);

jest.mock('@notifee/react-native', () =>
    require('@notifee/react-native/jest-mock')
);

export const mockOnTokenRefresh = jest.fn();
jest.mock('@react-native-firebase/messaging', () => ({
    __esModule: true,
    default: () => ({ onTokenRefresh: mockOnTokenRefresh }),
    FirebaseMessagingTypes: jest.fn()
}));

export const mockNavigate = jest.fn();
export const mockPop = jest.fn();
export const mockGoBack = jest.fn();
export const mockReplace = jest.fn();
export const mockNavigationDispatch = jest.fn();
export const mockUseRoute = jest.fn().mockImplementation(() => ({
    params: {
        id: 1,
        unit: 'kg',
        code: '11111',
        url: 'https://www.google.com/'
    }
}));

jest.mock('@/services/notification', () => ({
    resetNotifications: jest.fn(),
    registerBackgroundNotificationsListener: jest.fn(),
    getFCMToken: jest.fn().mockImplementation(() => {
        return Promise.resolve({});
    }),
    triggerFirebaseForegroundNotification: jest.fn(),
    registerFirebaseMessageListener: jest.fn(),
    requestNotificationPermissions: jest.fn()
}));

jest.mock('@react-navigation/native', () => {
    const real = jest.requireActual('@react-navigation/native');
    return {
        ...real,
        useNavigation: jest.fn().mockImplementation(() => {
            return {
                canGoBack: jest.fn().mockImplementation(() => {
                    return true;
                }),
                goBack: mockGoBack,
                navigate: mockNavigate,
                pop: mockPop,
                replace: mockReplace,
                dispatch: mockNavigationDispatch
            };
        }),
        useRoute: mockUseRoute,
        useIsFocused: jest.fn().mockImplementation(() => ({
            isFocused: true
        })),
        useFocusEffect: jest.fn()
    };
});

export const mockToast = jest.fn();

jest.mock('react-native-toast-message', () => ({
    show: mockToast
}));

export const mockGetChannelDisplayImage = jest.fn();
jest.mock('@/services/chat', () => ({
    getChannelDisplayImage: mockGetChannelDisplayImage
}));

export const mockConnectUser = jest.fn(() => Promise.resolve());
export const mockDisconnectUser = jest.fn();
export const mockGetInstance = jest.fn(() => ({
    connectUser: mockConnectUser,
    disconnectUser: mockDisconnectUser
}));
jest.mock('stream-chat', () => ({
    StreamChat: {
        getInstance: mockGetInstance
    }
}));

jest.mock('@aws-sdk/client-s3', () => ({
    S3Client: jest.fn(() => ({
        send: jest.fn(() => Promise.resolve())
    })),
    PutObjectCommand: jest.fn().mockImplementation(() => {
        return {};
    })
}));

jest.mock('react-native-pdf', () => null);
