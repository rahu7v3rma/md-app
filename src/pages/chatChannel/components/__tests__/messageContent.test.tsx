import React from 'react';
import { Linking, View as mockView } from 'react-native';
import { act, create } from 'react-test-renderer';

import { mockChatClient } from '@/jestSetup';

import MessageContent from '../messageContent';

const mockMessageContext = {
    message: {
        user: {
            id: '1',
            name: 'John Doe'
        },
        text: 'https://example.com This is a test message',
        latest_reactions: [],
        created_at: '2023-10-23T12:34:56.789Z',
        attachments: [],
        reaction_counts: {
            like: 2
        }
    },
    reactions: [],
    handleToggleReaction: jest.fn()
};

jest.mock('stream-chat-react-native', () => {
    return {
        Gallery: mockView,
        Giphy: mockView,
        FileAttachmentGroup: mockView,
        Attachment: mockView,
        useChatContext: jest.fn().mockImplementation(() => ({
            client: mockChatClient
        })),
        useMessageContext: jest
            .fn()
            .mockImplementation(() => mockMessageContext),
        ThumbsUpReaction: jest.fn().mockImplementation(() => {
            return <></>;
        })
    };
});

jest.mock('react-native-config', () => ({
    API_BASE_URL: 'your_mocked_api_base_url'
}));

describe('MessageContent', () => {
    it('matches snapshot', () => {
        const tree = create(<MessageContent />);
        expect(tree).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        const tree = create(<MessageContent />);
        expect(tree.root).toBeTruthy();
    });

    it('opens the URL when a link in the message is pressed', async () => {
        const openURLSpy = jest
            .spyOn(Linking, 'openURL')
            .mockImplementation(() => Promise.resolve());

        let tree: any;
        await act(async () => {
            tree = create(<MessageContent />);
        });

        const linkElement = tree.root.findByProps({
            testID: 'message-link'
        });
        linkElement.props.onPress();

        expect(openURLSpy).toHaveBeenCalledWith('https://example.com');

        openURLSpy.mockRestore();
    });

    it('displays the user name correctly', () => {
        const tree = create(<MessageContent />);

        const userNameElement = tree.root.findByProps({
            testID: 'user-name'
        });

        expect(userNameElement.props.children).toBe('John Doe');
    });

    it('displays the message text correctly', async () => {
        const testMessage = 'This is a test message';
        mockMessageContext.message.text = testMessage;

        const tree = create(<MessageContent />);

        const messageTextElement = tree.root.findByProps({
            testID: 'message-text'
        });

        expect(messageTextElement.props.children).toEqual([testMessage]);
    });

    it('displays the number of likes correctly', () => {
        const tree = create(<MessageContent />);

        const likeCountElement = tree.root.findByProps({
            testID: 'like-count'
        });

        expect(likeCountElement.props.children).toBe(2);
    });

    it('opens the reactions list when the like icon is pressed', async () => {
        let tree: any;
        await act(async () => {
            tree = create(<MessageContent />);
        });
        const likeIconElement = tree.root.findByProps({
            testID: 'like-icon'
        });
        await act(() => likeIconElement.props.onPress());
        let reactionsList;
        try {
            reactionsList = tree.root.findByProps({
                testID: 'reactionsList'
            }).props;
        } catch (e) {}
        expect(reactionsList).toBeTruthy();
    });
});
