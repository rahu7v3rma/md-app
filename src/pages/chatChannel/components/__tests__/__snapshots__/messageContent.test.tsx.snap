// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageContent matches snapshot 1`] = `
Array [
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    transparent={true}
    visible={false}
  />,
  <View
    onLayout={[Function]}
    style={
      Object {
        "width": "83%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "flex-start",
          "backgroundColor": "#ECF8E6",
          "borderBottomEndRadius": 16,
          "borderBottomStartRadius": 16,
          "borderColor": "#000000",
          "borderStyle": "solid",
          "borderTopEndRadius": 16,
          "borderTopLeftRadius": 0,
          "flexDirection": "row",
          "marginVertical": 8,
          "opacity": 1,
          "padding": 8,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 1,
        }
      }
    >
      <View
        style={
          Object {
            "width": "15%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/images/default_avatar.jpg",
              }
            }
            style={
              Object {
                "borderRadius": 20,
                "height": 40,
                "width": 40,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "marginLeft": 10,
            "width": "80%",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            adjustsFontSizeToFit={true}
            style={
              Array [
                Object {
                  "fontFamily": "Poppins",
                },
                Object {
                  "width": "80%",
                },
                Object {
                  "color": "#4D4F4D",
                  "fontSize": 16,
                  "fontWeight": "600",
                  "lineHeight": 24,
                },
              ]
            }
            testID="user-name"
          >
            John Doe
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
                "position": "relative",
                "top": 5,
              }
            }
          >
            <Text
              adjustsFontSizeToFit={true}
              style={
                Array [
                  Object {
                    "fontFamily": "Poppins",
                  },
                  Object {
                    "marginRight": 2,
                  },
                  Object {
                    "color": "#A4AAAF",
                    "fontSize": 12,
                    "fontWeight": undefined,
                    "lineHeight": 20,
                  },
                ]
              }
            >
              18:04
            </Text>
          </View>
        </View>
        <Text
          adjustsFontSizeToFit={true}
          style={
            Array [
              Object {
                "fontFamily": "Poppins",
              },
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "#271A51",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 22,
              },
            ]
          }
          testID="message-text"
        >
          <Text
            adjustsFontSizeToFit={true}
            onPress={[Function]}
            style={
              Array [
                Object {
                  "fontFamily": "Poppins",
                },
                Object {
                  "textDecorationColor": "#3ED1FF",
                  "textDecorationLine": "underline",
                },
                Object {
                  "color": "#3ED1FF",
                  "fontSize": undefined,
                  "fontWeight": undefined,
                  "lineHeight": undefined,
                },
              ]
            }
            testID="message-link"
          >
            https://example.com
          </Text>
           This is a test message
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ececec",
          "borderRadius": 20,
          "bottom": 0,
          "display": "flex",
          "flexDirection": "row",
          "opacity": 1,
          "paddingHorizontal": 7,
          "paddingVertical": 2,
          "position": "absolute",
          "right": -10,
          "zIndex": 1,
        }
      }
      testID="like-icon"
    >
      <Text
        adjustsFontSizeToFit={true}
        style={
          Array [
            Object {
              "fontFamily": "Poppins",
            },
            Object {
              "marginLeft": 1,
              "marginTop": 2,
            },
            Object {
              "color": "#271A51",
              "fontSize": 12,
              "fontWeight": "500",
              "lineHeight": 20,
            },
          ]
        }
        testID="like-count"
      >
        2
      </Text>
    </View>
  </View>,
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  />,
]
`;
